sql/schemachanger: Add support for storing policy command and type

spilchen opened the pull request:
A previous commit introduced basic support for CREATE/DROP POLICY. This commit expands on that functionality by storing additional details in the policy descriptor. Specifically, it adds support for storing the policy type (restrictive or permissive) and the policy command (ALL, SELECT, INSERT, UPDATE, or DELETE).

Since neither the policy type nor the policy command will be modifiable via ALTER POLICY, these attributes are included in the Policy element within the DSC, rather than as separate elements.

Epic: CRDB-11724
Informs: #136696
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138909)
<!-- Reviewable:end -->

Dedej-Bergin reviewed (commented):
Nice Work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138909#-:-OGVCHmL6d9-866BGBpO:bdcbtrm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138909#-OGVCPTu31GdUbMNT0CD:-OGVCPTu31GdUbMNT0CE:b-s2gbhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b30a2adeae7acdc833a5006180ecfcc74cc5c907/pkg/sql/logictest/testdata/logic_test/row_level_security#L125)):*
> ```
> 
> statement ok
> CREATE POLICY "policy 4" ON multi_pol_tab1 FOR INSERT
> ```

Why do we have 2 policies for `INSERT`, policy 4 and policy 6?
___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 140 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138909#-OGVCYzsE9XG4pkrNulQ:-OGVCYzsE9XG4pkrNulR:b-x9jnos) ([raw file](https://github.com/cockroachdb/cockroach/blob/b30a2adeae7acdc833a5006180ecfcc74cc5c907/pkg/sql/logictest/testdata/logic_test/row_level_security#L140)):*
> ```
> 
> query TT
> SHOW CREATE TABLE multi_pol_tab1
> ```

What are we trying to test with the `SHOW CREATE` query? I'm not seeing anything to do with policies in the query result.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138909)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/catalog/catpb/enum.proto` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138909#-OGVhDWk4Onpk9D34Iev:-OGVhDWk4Onpk9D34Iew:bpfzg07) ([raw file](https://github.com/cockroachdb/cockroach/blob/b30a2adeae7acdc833a5006180ecfcc74cc5c907/pkg/sql/catalog/catpb/enum.proto#L53)):*
> ```proto
> 
> // Type indicates the type of policy it is.
> enum PolicyType {
> ```

nit: can we make these enums implement `redact.SafeValue`? (we need to do that in some other file in the catpb package)
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 2084 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138909#-OGVhKT200zvxqvipQbb:-OGVhKT200zvxqvipQbc:b-sufq58) ([raw file](https://github.com/cockroachdb/cockroach/blob/b30a2adeae7acdc833a5006180ecfcc74cc5c907/pkg/sql/catalog/tabledesc/validate.go#L2084)):*
> ```Go
> 		}
> 		idToName[p.ID] = p.Name
> 		if _, ok := catpb.PolicyType_name[int32(p.Type)]; !ok || p.Type == catpb.PolicyType_POLICYTYPE_UNUSED {
> ```

is this validation potentially overly sensitive? the situation i wonder about is if we ever add a new enum value for either of these types, then use it in a mixed version cluster. the nodes on the older version will hit this validation, and will consider the descriptor corrupt as a result.

perhaps we could address this concern by adding a comment where the enums are defined that says that any new values that are added need to be gated behind a cluster version finalized check.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138909)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin)

___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138909#-OGVCPTu31GdUbMNT0CD:-OGVi1EK7V3XSxkCmswE:bkscxog) ([raw file](https://github.com/cockroachdb/cockroach/blob/b30a2adeae7acdc833a5006180ecfcc74cc5c907/pkg/sql/logictest/testdata/logic_test/row_level_security#L125)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

Why do we have 2 policies for `INSERT`, policy 4 and policy 6?
</blockquote></details>

Good point. There is no reason. Let me remove one.
___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 140 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138909#-OGVCYzsE9XG4pkrNulQ:-OGVhRmNBTPgjN5yDQE9:b-90g03r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b30a2adeae7acdc833a5006180ecfcc74cc5c907/pkg/sql/logictest/testdata/logic_test/row_level_security#L140)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

What are we trying to test with the `SHOW CREATE` query? I'm not seeing anything to do with policies in the query result.
</blockquote></details>

Eventually, `SHOW CREATE` will include the DDL for policies. Adding this here so that we don't forget to verify the output when a table has multiple policies.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138909)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin and @rafiss)

___
*[`pkg/sql/catalog/catpb/enum.proto` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138909#-OGVhDWk4Onpk9D34Iev:-OGVlkrs2QUoZPPqA5rK:bv93ow0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b30a2adeae7acdc833a5006180ecfcc74cc5c907/pkg/sql/catalog/catpb/enum.proto#L53)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: can we make these enums implement `redact.SafeValue`? (we need to do that in some other file in the catpb package)
</blockquote></details>

Good idea. I added `pkg/sql/catalog/catpb/redact.go`.
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 2084 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138909#-OGVhKT200zvxqvipQbb:-OGVluvg2zkHkrJd7EGz:b-4joj7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b30a2adeae7acdc833a5006180ecfcc74cc5c907/pkg/sql/catalog/tabledesc/validate.go#L2084)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is this validation potentially overly sensitive? the situation i wonder about is if we ever add a new enum value for either of these types, then use it in a mixed version cluster. the nodes on the older version will hit this validation, and will consider the descriptor corrupt as a result.

perhaps we could address this concern by adding a comment where the enums are defined that says that any new values that are added need to be gated behind a cluster version finalized check.
</blockquote></details>

I've added the comment. However, is there an existing precedent for verifying enum values?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138909)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin and @spilchen)

___
*[`pkg/sql/catalog/tabledesc/validate.go` line 2084 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138909#-OGVhKT200zvxqvipQbb:-OGVtM2BDEGBrhTNiNIR:b43vx5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/b30a2adeae7acdc833a5006180ecfcc74cc5c907/pkg/sql/catalog/tabledesc/validate.go#L2084)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I've added the comment. However, is there an existing precedent for verifying enum values?
</blockquote></details>

i believe in many cases we don't actually validate enum values so that we can easily add more later. however, in this case, if we never expect to add other values in the future, it makes sense to have the validation.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: Add support for storing policy command and type

A previous commit introduced basic support for CREATE/DROP POLICY. This
commit expands on that functionality by storing additional details in
the policy descriptor. Specifically, it adds support for storing the
policy type (restrictive or permissive) and the policy command (ALL,
SELECT, INSERT, UPDATE, or DELETE).

Since neither the policy type nor the policy command will be modifiable
via ALTER POLICY, these attributes are included in the Policy element
within the DSC, rather than as separate elements.

Epic: CRDB-11724
Informs: #136696
Release note: None

Dedej-Bergin reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138909#-:-OGWLZJFDJoVbBwNWgFG:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
